microprofile/openapi/src/main/java/io/helidon/microprofile/openapi/IndexBuilder.java

```diff
@@ -0,0 +1,159 @@
+
+    private static final String INDEX_PATH = "/META-INF/jandex.idx";
+
+    private static final Logger LOGGER = Logger.getLogger(IndexBuilder.class.getName());
+
+    private final boolean isIndexPresentOnClasspath;
+
+    private final Set<Class<?>> annotatedTypes = new HashSet<>();
+
+    /**
+     * Creates a new instance of the index builder.
+     *
+     * @throws IOException in case of error checking for the Jandex index file
+     */
+    public IndexBuilder() throws IOException {
+        isIndexPresentOnClasspath = checkForIndexFile();
+        if (isIndexPresentOnClasspath) {
+            LOGGER.log(Level.FINE, () -> String.format("Index file %s was located and will be used", INDEX_PATH));
+        } else {
+            LOGGER.log(Level.INFO, () -> String.format(
+                    "OpenAPI support could not locate the index file %s "
+                    + "so will build an in-memory index. This slows your app start-up.%n"
+                    + "Consider using the jandex maven plug-in to build the index "
+                    + "and add it to your app at build-time",
+                    INDEX_PATH));
+        }
+    }
+
+    /**
+     * Records each type that is annotated unless a Jandex index was found on
+     * the classpath (in which case we do not need to build our own in memory).
+     *
+     * @param <X> annotated type
+     * @param event {@code ProcessAnnotatedType} event
+     */
+    <X> void processAnnotatedType(@Observes ProcessAnnotatedType<X> type) {
+        if (isIndexPresentOnClasspath) {
+            return;
+        }
+        Class<?> c = type.getAnnotatedType().getJavaClass();
+        annotatedTypes.add(c);
+    }
+
+    /**
+     * Reports an {@link IndexView} for the Jandex index that describes
+     * annotated classes for endpoints.
+     *
+     * @return {@code IndexView} describing discovered classes
+     * @throws IOException in case of error reading an existing index file or
+     * reading class bytecode from the classpath
+     */
+    public IndexView indexView() throws IOException {
+        return isIndexPresentOnClasspath ? existingIndexFileReader() : indexFromHarvestedClasses();
+    }
+
+    private IndexView existingIndexFileReader() throws IOException {
+        try (InputStream jandexIS = getClass().getResourceAsStream(INDEX_PATH)) {
+            if (jandexIS == null) {
+                throw new IllegalArgumentException("Attempted to read from previously-located index file "
+                        + INDEX_PATH + " but the file cannot be found");
+            }
+            LOGGER.log(Level.FINE, "Using Jandex index at {0}", INDEX_PATH);
+            return new IndexReader(jandexIS).read();
+        }
+    }
+
+    private IndexView indexFromHarvestedClasses() throws IOException {
+        Indexer indexer = new Indexer();
+        for (Class<?> c : annotatedTypes) {
+            try (InputStream is = contextClassLoader().getResourceAsStream(resourceNameForClass(c))) {
```

Comment:
Consider `BufferedInputStream`.

Reply:
The invoked method does its own buffering.

Reply:
Oh!  You mean `URL#openStream()` (called by `ClassLoader#getResourceAsStream()`) is guaranteed to buffer?  I didn't know that!  Learn something new every day.